File: Code/C++/1140. Stone Game II.cpp
#include <vector>
#include <numeric>
#include <algorithm>
#include <cstring>

using namespace std;

#include <vector>
#include <numeric>
#include <algorithm>
#include <cstring>

using namespace std;

class Solution {
public:
    int stoneGameII(vector<int>& piles) {
        int n = piles.size();
        vector<int> suffix(n+1);
        vector<vector<int>> dp(n+1, vector<int>(n+1, -1));
        for (int i = n-1; i>=0; i--){
            suffix[i] = suffix[i+1] + piles[i];
        }
        return dfs(0, 1, suffix, dp);

    }

    int dfs(int i, int m, vector<int> &suffix, vector<vector<int>> &dp){
        if (i >= suffix.size() - 1) {return 0;}
        if (dp[i][m] != -1) {return dp[i][m];}
        int res = 0;
        for (int x=1; x<=2*m; x++){
            if (i + x <= suffix.size() - 1){
                res = max(res, suffix[i] - dfs(i+x, max(x, m), suffix, dp));
            }
        }
        dp[i][m] = res;
        return res;
    }

};

File: Code/Python/1140. Stone Game II.py
from itertools import accumulate
from functools import cache

class Solution:
    def stoneGameII(self, piles) -> int:
        # xx | xxxx | xxxx..
        #    i      j            
        n = len(piles)
        suffix = list(accumulate(piles[::-1]))[::-1]
        # dfs calc max stones Alice can get from index i with the constraint M
        @cache
        def dfs(i, M):
            if i >= n:
                return 0
            res = 0 
            for x in range(1, 2*M + 1):
                if i + x <= n:
                    res = max(res, suffix[i] - dfs(i+x, max(x, M)))
            return res
        
        return dfs(0, 1)

